File: skill_doctor/report/sarif.py
"""
SARIF 2.1 report generation.
"""

import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

from skill_doctor.models import ScanResult


def generate_sarif(result: ScanResult, output_path: Optional[Path] = None) -> str:
    """
    Generate SARIF 2.1 report from scan result.

    Args:
        result: ScanResult to convert to SARIF
        output_path: Optional path to write SARIF file

    Returns:
        SARIF JSON string
    """
    sarif = {
        "version": "2.1.0",
        "$schema": "https://json.schemastore.org/sarif-2.1.0.json",
        "runs": [
            {
                "tool": {
                    "driver": {
                        "name": "Skill Doctor",
                        "version": "0.1.0",
                        "informationUri": "https://github.com/KalarisLabs/Skill-Doctor",
                        "rules": [
                            {
                                "id": f.category,
                                "name": f.category,
                                "shortDescription": {"text": f.description},
                                "help": {"text": f.remediation},
                            }
                            for f in result.findings
                        ],
                    }
                },
                "invocations": [
                    {
                        "startTimeUtc": result.scanned_at.isoformat(),
                        "endTimeUtc": datetime.now(timezone.utc).isoformat(),
                        "exitCode": 0 if result.risk_level == "SAFE" else 1,
                    }
                ],
                "results": [
                    {
                        "ruleId": finding.category,
                        "level": _map_severity_to_level(finding.severity),
                        "message": {"text": finding.description},
                        "locations": [
                            {
                                "physicalLocation": {
                                    "artifactLocation": {"uri": finding.file},
                                    "region": (
                                        {"startLine": finding.line} if finding.line else {}
                                    ),
                                }
                            }
                        ],
                        "properties": {
                            "confidence": finding.confidence,
                            "engine": finding.engine,
                        },
                    }
                    for finding in result.findings
                ],
            }
        ],
    }

    sarif_json = json.dumps(sarif, indent=2)

    if output_path:
        output_path.write_text(sarif_json, encoding="utf-8")

    return sarif_json


def _map_severity_to_level(severity: str) -> str:
    """Map Finding severity to SARIF level."""
    mapping = {
        "CRITICAL": "error",
        "HIGH": "error",
        "MEDIUM": "warning",
        "LOW": "note",
        "INFO": "note",
    }
    return mapping.get(severity, "note")

File: crates/report/src/sarif.rs
//! SARIF 2.1 report generation for GitHub Code Scanning integration.

use std::path::Path;

use anyhow::Result;
use serde_json::json;
use skill_doctor_core::models::ScanResult;

/// Generate a SARIF 2.1.0 report file from scan results.
pub fn generate_sarif(result: &ScanResult, output_path: &Path) -> Result<()> {
    let rules: Vec<serde_json::Value> = result
        .findings
        .iter()
        .map(|f| {
            json!({
                "id": f.category.replace(" · ", "_").replace(' ', "_"),
                "name": f.category,
                "shortDescription": { "text": f.description },
                "defaultConfiguration": {
                    "level": match f.severity {
                        skill_doctor_core::Severity::Critical | skill_doctor_core::Severity::High => "error",
                        skill_doctor_core::Severity::Medium => "warning",
                        _ => "note",
                    }
                }
            })
        })
        .collect();

    let results: Vec<serde_json::Value> = result
        .findings
        .iter()
        .map(|f| {
            json!({
                "ruleId": f.category.replace(" · ", "_").replace(' ', "_"),
                "level": match f.severity {
                    skill_doctor_core::Severity::Critical | skill_doctor_core::Severity::High => "error",
                    skill_doctor_core::Severity::Medium => "warning",
                    _ => "note",
                },
                "message": { "text": f.description },
                "locations": [{
                    "physicalLocation": {
                        "artifactLocation": { "uri": f.file },
                        "region": {
                            "startLine": f.line.unwrap_or(1)
                        }
                    }
                }],
                "properties": {
                    "confidence": f.confidence,
                    "engine": format!("{}", f.engine),
                    "remediation": f.remediation,
                }
            })
        })
        .collect();

    let sarif = json!({
        "$schema": "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/main/sarif-2.1/schema/sarif-schema-2.1.0.json",
        "version": "2.1.0",
        "runs": [{
            "tool": {
                "driver": {
                    "name": "Skill Doctor",
                    "version": env!("CARGO_PKG_VERSION"),
                    "informationUri": "https://github.com/KalarisLabs/Skill-Doctor",
                    "rules": rules,
                }
            },
            "results": results,
        }]
    });

    let json_str = serde_json::to_string_pretty(&sarif)?;
    std::fs::write(output_path, json_str)?;
    Ok(())
}
